crates/sui-json-rpc/src/coin_api.rs

```diff
@@ -288,41 +210,716 @@ impl CoinReadApiServer for CoinReadApi {
+        &self,
+        digest: TransactionDigest,
+    ) -> SuiResult<(VerifiedTransaction, TransactionEffects)> {
+        self.state
+            .get_executed_transaction_and_effects(digest)
+            .await
+    }
+
+    async fn get_balance(&self, owner: SuiAddress, coin_type: TypeTag) -> SuiResult<TotalBalance> {
+        self.state
+            .indexes
+            .as_ref()
+            .ok_or(SuiError::IndexStoreNotAvailable)?
+            .get_balance(owner, coin_type)
+            .await
+    }
+
+    async fn get_all_balance(
+        &self,
+        owner: SuiAddress,
+    ) -> SuiResult<Arc<HashMap<TypeTag, TotalBalance>>> {
+        self.state
+            .indexes
+            .as_ref()
+            .ok_or(SuiError::IndexStoreNotAvailable)?
+            .get_all_balance(owner)
+            .await
+    }
+}
+
+#[cached(
+    type = "SizedCache<String, ObjectID>",
+    create = "{ SizedCache::with_size(10000) }",
+    convert = r#"{ format!("{}{}", package_id, object_struct_tag) }"#,
+    result = true
+)]
+async fn find_package_object_id(
+    state: Arc<AuthorityState>,
+    package_id: ObjectID,
+    object_struct_tag: StructTag,
+) -> SuiApiResult<ObjectID> {
+    spawn_monitored_task!(async move {
+        let publish_txn_digest = state.find_publish_txn_digest(package_id)?;
+
+        let (_, effect) = state
+            .get_executed_transaction_and_effects(publish_txn_digest)
+            .await?;
+
+        for ((id, _, _), _) in effect.created() {
+            if let Ok(object_read) = state.get_object_read(id) {
+                if let Ok(object) = object_read.into_object() {
+                    if matches!(object.type_(), Some(type_) if type_.is(&object_struct_tag)) {
+                        return Ok(*id);
+                    }
+                }
+            }
+        }
+        Err(SuiRpcInputError::GenericNotFound(format!(
+            "Cannot find object [{}] from [{}] package event.",
+            object_struct_tag, package_id,
+        ))
+        .into())
+    })
+    .await?
+}
+
+/// CoinReadInternal trait to capture logic of interactions with AuthorityState and metrics
+/// This allows us to also mock internal implementation for testing
+#[cfg_attr(test, automock)]
+#[async_trait]
+pub trait CoinReadInternal {
+    fn get_state(&self) -> Arc<dyn State + Send + Sync>;
+    async fn get_coins_iterator(
+        &self,
+        owner: SuiAddress,
+        cursor: (String, ObjectID),
+        limit: Option<usize>,
+        one_coin_type_only: bool,
+    ) -> SuiApiResult<CoinPage>;
+    async fn get_coin_metadata(
+        &self,
+        coin_struct: StructTag,
+    ) -> SuiApiResult<Option<SuiCoinMetadata>>;
+    async fn get_treasury_cap(&self, coin_struct: StructTag) -> SuiApiResult<TreasuryCap>;
+}
+
+pub struct CoinReadInternalImpl {
+    // Trait object w/ Arc as we have methods that require sharing this across multiple threads
+    state: Arc<dyn State + Send + Sync>,
```

Comment:
Here, we can either accept an `Arc<dyn State ...>` and implement `State` directly on `AuthorityState`, or we can accept a `Box<dyn State ...>` and use the wrapper type, but we shouldn't do both.

Reply:
I find that the function find_package_by_object_id is throwing a wrench into either approaches - it requires an Arc<AuthorityState> input, so if I implement State for AuthorityState, I'm unable to pass Arc<AuthorityState> as self is &AuthorityState

And due to the function spawning a tokio task I think we need the arc, but I could be wrong here

Perhaps the correct thing to do is to lift find_package_object into CoinReadImpl.

Reply:
Was there an issue with changing the parameter for `find_package_by_object_id` to accept an `Arc<dyn State...>` as well?

Reply:
Decided on implementing `State` for `AuthorityState`, thanks Ashok